src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs

```diff
@@ -0,0 +1,302 @@
+            _projectFileModelWatcher = _projectFileWatcherFactory.CreateExport();
+            _projectFileModelWatcher.Initialize();
+
+            // Finally, show the frame and move to EditorClean in lockstep
+            lock (_lock)
+            {
+                _windowFrame.Show();
+                _currentState = EditorState.EditorClean;
+                _projectFileModelWatcher.Initialize();
+            }
+        }
+
+        /// <summary>
+        /// Called from the editor wrapper. This sets up the listener that controls the dirty state and saving commands.
+        /// </summary>
+        public async Task InitializeTextBufferStateListenerAsync(WindowPane hostPane)
+        {
+            _textBufferStateListener = _textBufferListenerFactory.CreateExport();
+            await _textBufferStateListener.InitializeAsync(hostPane).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Starts the process of closing the project file editor, if one is open currently.
+        /// </summary>
+        public async Task CloseWindowAsync()
+        {
+            lock (_lock)
+            {
+                if (_currentState == EditorState.NoEditor) return;
+
+                // If there was no changes, then we just close the editor
+                if (_currentState == EditorState.EditorClean || _currentState == EditorState.ProjectFileChangingFromClean)
+                {
+                    _currentState = EditorState.EditorClosing;
+                }
+
+                Requires.Range(_currentState != EditorState.EditorDirty && _currentState != EditorState.ProjectFileChangingFromDirty, nameof(_currentState));
+            }
+
+            _projectFileModelWatcher?.Dispose();
+            _textBufferStateListener?.Dispose();
+            await (_frameEventsListener?.DisposeAsync() ?? Task.CompletedTask).ConfigureAwait(false);
+            _projectFileModelWatcher = null;
+            _frameEventsListener = null;
+            _textBufferStateListener = null;
+
+            lock (_lock)
+            {
+                _currentState = EditorState.NoEditor;
+            }
+            return;
+        }
+
+        #endregion
+
+        #region Update Project File
+
+        /// <summary>
+        /// Schedules the project file to be updated. This is called by the ProjectFileModelWatcher, which is within a Project write lock.
+        /// Because of the time-sensitive nature of that lock, instead of updating the file immediately we schedule an update as soon
+        /// as the JTF can run our code.
+        /// </summary>
+        public void ScheduleProjectFileUpdate()
+        {
+            lock (_lock)
+            {
+                _currentState = _currentState == EditorState.EditorDirty ?
+                    EditorState.ProjectFileUpdateScheduledFromDirty : EditorState.ProjectFileUpdateScheduledFromClean;
+                _threadingService.JoinableTaskFactory.RunAsync(UpdateProjectFileAsync);
```

Comment:
Where are exceptions from this observed? What's the default behavior of JTF on unobserved exceptions?